junit-vintage-engine/src/test/java/org/junit/vintage/engine/execution/ParallelExecutionIntegrationTests.java

```diff
@@ -67,6 +75,57 @@ void executesTestClassesInParallel(TestReporter reporter) {
 		assertThat(threadNames).hasSize(3);
 	}
 
+	@Test
+	void executesTestMethodsInParallel(TestReporter reporter) {
+		JUnit4ParallelMethodsTestCase.AbstractBlockingTestCase.threadNames.clear();
+		JUnit4ParallelMethodsTestCase.AbstractBlockingTestCase.countDownLatch = new CountDownLatch(3);
+
+		var events = executeInParallelSuccessfully(3, false, true, FirstMethodTestCase.class).list();
+
+		var startedTimestamps = getTimestampsFor(events, event(test(SEGMENT_TYPE_TEST), started()));
+		var finishedTimestamps = getTimestampsFor(events, event(test(SEGMENT_TYPE_TEST), finishedSuccessfully()));
+		var threadNames = new HashSet<>(JUnit4ParallelMethodsTestCase.AbstractBlockingTestCase.threadNames);
+
+		reporter.publishEntry("startedTimestamps", startedTimestamps.toString());
+		reporter.publishEntry("finishedTimestamps", finishedTimestamps.toString());
+
+		assertThat(startedTimestamps).hasSize(3);
+		assertThat(finishedTimestamps).hasSize(3);
+		assertThat(startedTimestamps).allMatch(startTimestamp -> finishedTimestamps.stream().noneMatch(
+			finishedTimestamp -> finishedTimestamp.isBefore(startTimestamp)));
+		assertThat(threadNames).hasSize(3);
+	}
+
+	@Test
+	void executesTestClassesAndMethodsInParallel(TestReporter reporter) {
+		JUnit4ParallelMethodsTestCase.AbstractBlockingTestCase.threadNames.clear();
+		JUnit4ParallelMethodsTestCase.AbstractBlockingTestCase.countDownLatch = new CountDownLatch(9);
+
+		var events = executeInParallelSuccessfully(3, true, true, FirstMethodTestCase.class, SecondMethodTestCase.class,
+			ThirdMethodTestCase.class).list();
+
+		var startedClassesTimestamps = getTimestampsFor(events, event(container(SEGMENT_TYPE_RUNNER), started()));
+		var finishedClassesTimestamps = getTimestampsFor(events,
+			event(container(SEGMENT_TYPE_RUNNER), finishedSuccessfully()));
+		var startedMethodsTimestamps = getTimestampsFor(events, event(test(SEGMENT_TYPE_TEST), started()));
+		var finishedMethodsTimestamps = getTimestampsFor(events,
+			event(test(SEGMENT_TYPE_TEST), finishedSuccessfully()));
+
+		var threadNames = new HashSet<>(JUnit4ParallelMethodsTestCase.AbstractBlockingTestCase.threadNames);
+
+		reporter.publishEntry("startedClassesTimestamps", startedClassesTimestamps.toString());
+		reporter.publishEntry("finishedClassesTimestamps", finishedClassesTimestamps.toString());
+		reporter.publishEntry("startedMethodsTimestamps", startedMethodsTimestamps.toString());
+		reporter.publishEntry("finishedMethodsTimestamps", finishedMethodsTimestamps.toString());
+
+		assertThat(startedClassesTimestamps).hasSize(3);
+		assertThat(finishedClassesTimestamps).hasSize(3);
+		assertThat(startedMethodsTimestamps).hasSize(9);
+		assertThat(finishedMethodsTimestamps).hasSize(9);
+
+		assertThat(threadNames).hasSizeGreaterThanOrEqualTo(2);
```

Comment:
I think the problem here is that the `ForkJoinPool` is not stealing work when we're waiting for the `RunnerDescriptors` children (see next comment).

```suggestion
		assertThat(threadNames).hasSize(3);
```

Reply:
I did notice that the `ForkJoinPool` isn't stealing work, but I didn't realize that it was because of waiting for the `RunnerDescriptors`' children.  

Running the Futures separately seems like a great approach.  

I learned a lot from this—thank you so much! 🙇🏻‍♂️